### DISCUSSION TITLE
XKB multilevel keyboard support

### DISCUSSION THREAD

**Discussion by dsc8x** [qtile/qtile#2918] [Ideas]
Right now, only the first two levels of a keyboard layout are supported for key bindings and widget text input (normal level and shift level). But there are some keyboard layouts out there, which are using up to 8 levels, as for example:

http://neo-layout.org/
https://github.com/epegzz/coder-layout

Note, that both of them have the number keys on a higher level instead of (only) on the usual numbers row on the first level.
The second layout does not even have numbers on the first 2 levels, making it impossible to type numbers in qtile.

The root cause of this problem: qtile only captures key codes, not the translated symbols. But on multilevel keyboard layouts, one key code may generate several different symbols.

I know, that xcb does not offer any way to capture the XLookupString value right now, but I have an idea how to make multilevel keyboard layouts work with qtile nevertheless.

Pull request might follow :-)

**Comment by tych0** [comment]
Pull request would definitely be welcome. I think we could use some more work in this area, actually. I've never particularly understood how all the XKB stuff works (which is how we ended up with hacks like 43972389a7d36ff06d2e639b8d0350d6016c639d).

**Comment by dequis** [comment]
Also I'd love to be able to use extended input methods through XIM - if xterm can do it, we should be able to do so too, right?

(I like to consider xlib a "subset" of xcb, since the former is actually implemented using the latter nowadays.)

**Comment by eigengrau** [comment]
This problem doesn’t seem to be about the levels, really. I have this with the AdWN keyboard layout, but the same should apply to Neo. Many of the keys not working for me are actually on level one, such as the German umlauts or the numeric keypad. Someone more knowlegeable about keyboard layouts related to me once that this behavior is indicative that lock the modifiers are ignored:

https://groups.google.com/forum/#!topic/adnw/dqroq3lVeys (German)

The posting was related to the Gnome shell, which has this problem when using XIM. But maybe the root cause is related. HTH.
